src/Microsoft.AspNet.Server.Kestrel/Http/ListenerSecondary.cs

```diff
@@ -29,95 +31,112 @@ protected ListenerSecondary(ServiceContext serviceContext) : base(serviceContext
-                                        }
-
-                                        if (DispatchPipe.PendingCount() == 0)
-                                        {
-                                            return;
-                                        }
-
-                                        var acceptSocket = CreateAcceptSocket();
-
-                                        try
-                                        {
-                                            DispatchPipe.Accept(acceptSocket);
-                                        }
-                                        catch (UvException ex)
-                                        {
-                                            Log.LogError("DispatchPipe.Accept", ex);
-                                            acceptSocket.Dispose();
-                                            return;
-                                        }
-
-                                        var connection = new Connection(this, acceptSocket);
-                                        connection.Start();
-                                    },
-                                    null);
-
-                                tcs.SetResult(0);
-                            }
-                            catch (Exception ex)
-                            {
-                                DispatchPipe.Dispose();
-                                tcs.SetException(ex);
-                            }
-                        },
-                        null);
-                }
-                catch (Exception ex)
+                DispatchPipe.Init(Thread.Loop, true);
+                var connect = new UvConnectRequest(Log);
+                connect.Init(Thread.Loop);
+                connect.Connect(
+                    DispatchPipe,
+                    _pipeName,
+                    (connect2, status, error, state) => ConnectCallback(connect2, status, error, (TaskCompletionSource<int>)state),
+                    tcs);
+            }
+            catch (Exception ex)
+            {
+                DispatchPipe.Dispose();
+                tcs.SetException(ex);
+            }
+        }
+
+        private static void ConnectCallback(UvConnectRequest connect, int status, Exception error, TaskCompletionSource<int> tcs)
+        {
+            var listener = (ListenerSecondary)tcs.Task.AsyncState;
+            listener.ConnectedCallback(connect, status, error, tcs);
+        }
+        private void ConnectedCallback(UvConnectRequest connect, int status, Exception error, TaskCompletionSource<int> tcs)
+        {
+            connect.Dispose();
+            if (error != null)
+            {
+                tcs.SetException(error);
+                return;
+            }
+
+            try
+            {
+                var ptr = Marshal.AllocHGlobal(4);
+                var buf = Thread.Loop.Libuv.buf_init(ptr, 4);
+
+                DispatchPipe.ReadStart(
```

Comment:
Both of these callbacks result in closure allocation because of buf and ptr. This is probably one of the more important places to avoid it, since it is per connection. Since we don't actually use the buffer, I think we could make the pointer and buffer a field, and only free it when ListenerSecondary is disposed.

What do you think?